### PR TITLE
Improve Tex string splitting at double braces: only split for double brace groups

### LINKED ISSUE #1244: Tex parsing issue (introduced in 0.5.0 in break_up_tex_strings)
## Description of bug / unexpected behavior
d = MathTex(r"\frac{dE}{d\mathbf{w}} =")

This produces an incorrect result. See below

TexFractionWithBoldFaceBug_ManimCE_v0 5 0

## Expected behavior

This bug was not present in 0.4.0

## How to reproduce the issue

```py
class TexFractionWithBoldFaceBug(Scene):
    def construct(self):
        d1 = MathTex(r"\frac{dE}{d\mathbf{w}}", "=")
        self.add(d1)
```

## System specifications

<details><summary>System Details</summary>

- OS (with version, e.g Windows 10 v2004 or macOS 10.15 (Catalina)): macOS 11
- RAM:
- Python version (`python/py/python3 --version`):3.7
- Installed modules (provide output from `pip list`):
```
PASTE HERE
```
</details>

<details><summary>LaTeX details</summary>

+ LaTeX distribution (e.g. TeX Live 2020):
+ Installed LaTeX packages:

</details>

<details><summary>FFMPEG</summary>

Output of `ffmpeg -version`:

```
PASTE HERE
```
</details>

## Additional comments

I have debugged and isolated the portion of the code that has introduced the issue.

break_up_tex_strings has been changed in 0.5.0; essentially borrowed from original manim

```
def break_up_tex_strings(self, tex_strings):
        print("input -> ", tex_strings)
        tex_strings = [str(t) for t in tex_strings]
        # Separate out anything surrounded in double braces
        patterns = ["{{", "}}"]
        # Separate out any strings specified in the isolate
        # or tex_to_color_map lists.
        patterns.extend([
            "({})".format(re.escape(ss)) for ss in it.chain(
                self.substrings_to_isolate, self.tex_to_color_map.keys())
        ])
        pattern = "|".join(patterns)
        pieces = []
        for s in tex_strings:
            pieces.extend(re.split(pattern, s))
        r = list(filter(lambda s: s, pieces))
        print("result -> ", r)
        return r
```

Here is the snapshot of what this new implementation produces with my print statements. Rest is identical to master and 0.5.0.

break_up_tex_string_log

As you can see that the returned value is malformed and is also very clear why this would happen. Essentially `patterns = ["{{", "}}"]` would result into this.

Just for testing if you change `patterns = ["{{", "}}"]` to `patterns = ["{{"]`, the bug will go away.

Another workaround is to construct the tex as `MathTex(r"\frac{dE}{d\mathbf{w} }", "=")`.  Note that I have put an extra space between the two closing brackets i.e. "} }" instead of "}}"

I use Latex a lot and based on my experience I envision that this is not the only time when two curly braces will appear.

### PR DESCRIPTION
**Pull request by behackl**
## Motivation

As a convenience method, #1037 introduced easier separation of substrings in Tex strings by putting parts that should be isolated in pairs of double braces. While the method itself is in fact convenient, the implementation was a bit too simple: it simply treated `{{` and `}}` as special delimiters at which the Tex string would be broken.

This PR improves the implementation by breaking the string only if a full double brace group is present.

## Testing Status

Added two doctests.

## Further Comments

Fixes #1244.

## Checklist
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have written a descriptive PR title (see top of PR template for examples)
- [x] I have written a changelog entry for the PR or deem it unnecessary
- [x] My new functions/classes either have a docstring or are private
- [x] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] My new documentation builds, looks correctly formatted, and adds no additional build warnings

## Reviewer Checklist
- [ ] The PR title is descriptive enough
- [ ] The PR is labeled correctly
- [ ] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

### CODE REVIEW

**Comment by behackl** [comment]
Just because I stumbled over it: 3b1b/manim seems to have removed the double brace notation altogether again recently (literally the most recent commit there). I still think that the notation is useful, and by enabling the split if and only if there is a full pair `{{` and `}}` in the string, most problems with it should be resolved automatically.

**Comment by ksachdeva** [comment]
The logic looks ok to me but as I mentioned in my reported issue the braces are kind of sacred in latex. And seems that 3b1b also thought so.

Though the logic of handling the braces ("}}") conditioned on if we encountered "{{" seems sound and works there are 3 more thoughts -

a) Given the role braces play in latex is it wise to overload them for functionality we desire?
b) Does it confuse the beginners in latex?
c) Is there a merit in having the same approach as 3b1b ... at least it should make the transfer of changes to and fro from his repo easier for future updates/mods/fixes etc

**Comment by newell** [comment]
I tested your change and it does fix the issue mentioned in the bug.  I agree with @ksachdeva that curly braces are somewhat sacred in LaTeX and also think the other concerns he raised are valid.  My personal view is that we shouldn't be using curly braces for this and either we should adopt another convention or follow what 3b1b is adopting.

**Comment by behackl** [comment]
There are not many situation where it makes sense in LaTeX that double opening braces occur naturally, and even less situations where double opening and closing braces occur naturally.

I agree with you in principle - if there were a better, similarly simple solution, I'd like to go for that -- but I'm not sure that there is another useful "inline" notation for this. If we are going to have some sort of inline notation, then I only really see braces as an option -- especially *because* they play a special role in LaTeX. Everything other than braces is being typeset -- and thus doesn't really allow for similarly simple workarounds like adding a space in between.

> b) Does it confuse the beginners in latex?

Currently: yes. With the solution proposed here: I don't really see how. I think you have to actively use this sort of notation -- but we could theoretically also be more explicit: if `MathTex` / `Tex` errors out after such an inline replacement has been made, we could raise an explicit error message explaining that {{ }} is special notation.

> c) Is there a merit in having the same approach as 3b1b ... at least it should make the transfer of changes to and fro from his repo easier for future updates/mods/fixes etc

I don't think that this should be a priority. Porting changes to/from 3b1b/manim already has a certain overhead as our repositories are strongly diverged. It's not like we can currently copy/paste code to exchange features, we already need to adapt the code generously.

**Comment by ksachdeva** [comment]
Unfortunately, there is no best solution to this problem so far; even if there was one, my recommendation would be to merge this PR as soon as possible as this first and foremost fixes the bug/regression introduced in 0.5.0

If possible, the alternative solution should be discussed as and when more user feedback arises and/or some other issues related to parsing show up.

**Comment by kilacoda** [comment]
Maybe we could change the splitting delimiter from `{{` to something less used in LaTeX, like `||`?

**Comment by behackl** [comment]
> Maybe we could change the splitting delimiter from `{{` to something less used in LaTeX, like `||`?

As mentioned above: if we have some inline notation at all, I think we should stick with the braces. If we go for `||` instead, users can't render `||` anymore -- which isn't the case with `{{` / `}}` because they have a special meaning in LaTeX.

The best solution from my point of view would be to raise a custom error if a string fails to compile after `break_up_tex_string` split the string.

**Review by jsonvillanueva — APPROVED** [review]
> The best solution from my point of view would be to raise a custom error if a string fails to compile after break_up_tex_string split the string.

I think the implementation is fine as is (though not perfect -- see example after this); would you rather raise this proposed error here in this PR, or later?

Example: `a^{ {{a + {{b}} }} }`
I think this input would break this implementation as `tex_strings` would get broken down into `["a^{ ", "a + {{b", " }} }"]`. An algorithm for correctness needs to consider matching brackets (and not using regular expressions would be more performant).

**Comment by behackl** [comment]
> > The best solution from my point of view would be to raise a custom error if a string fails to compile after break_up_tex_string split the string.
> 
> I think the implementation is fine as is (though not perfect -- see example after this); would you rather raise this proposed error here in this PR, or later?
> 
> Example: `a^{ {{b + {{c}} }} }` (not sure how realistic this is an example -- but I suppose people might want to color their variables/operators differently and split on them)
> I think this input would break this implementation as `tex_strings` would get broken down into `["a^{ ", "b + {{c", " }} }"]`. An algorithm for correctness needs to consider matching brackets (and not using regular expressions would be more performant).

I don't think your example would break: the internal implementation balances brackets before compiling, and here it should be able to repair all of the problems.

But anyhow, you still have a point of course. The unintuitive thing about the `{{ }}` notation is that it is not *meant* to be nested, there is no advantage in doing so and everything can be achieved by un-nested parenthesis as well.

**Comment by behackl** [comment]
I've added an explicit warning via 9b36a7b8ca30183f34cf10f96c259d061c38fba6. (I kinda hate how I had to implement that, but I don't see a lot of room for improvement given how convoluted all of the TeX-related code is.)

Example:

```py
>>> MathTex(r"\text{{ asdf }}")
[04/16/21 14:30:42] INFO     Writing "\text  asdf" to media/Tex/34c00056180becc8.tex                                                                                                                                   tex_file_writing.py:81
[04/16/21 14:30:43] INFO     Writing "\text" to media/Tex/9bb84ea3b159150a.tex                                                                                                                                         tex_file_writing.py:81
                    ERROR    LaTeX compilation error! latex reports:                                                                                                                                                  tex_file_writing.py:178
                    ERROR    ! Misplaced alignment tab character &.                                                                                                                                                   tex_file_writing.py:189
                                                                                                                                                                                                                                             
                    ERROR    \math@cr@@@ ->&                                                                                                                                                                          tex_file_writing.py:189
                                                                                                                                                                                                                                             
                    ERROR                   \omit \global \advance \row@ \@ne \ifst@rred \nonumber \fi \i...                                                                                                          tex_file_writing.py:189
                                                                                                                                                                                                                                             
                    ERROR    l.28 \end{align*}                                                                                                                                                                        tex_file_writing.py:189
                                                                                                                                                                                                                                             
                    ERROR    A group of double braces, {{ ... }}, was detected in your string.  Manim splits TeX strings at the double braces, which might have caused the current compilation error. If you didn't use    tex_mobject.py:437
                             the double brace split intentionally, add spaces between the braces to avoid the automatic splitting: {{ ... }} --> { { ... } }.                                                                                
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/behackl/code/manim/manim/mobject/svg/tex_mobject.py", line 444, in __init__
    raise compilation_error
  File "/Users/behackl/code/manim/manim/mobject/svg/tex_mobject.py", line 434, in __init__
    self.break_up_by_substrings()
  File "/Users/behackl/code/manim/manim/mobject/svg/tex_mobject.py", line 487, in break_up_by_substrings
    sub_tex_mob = SingleStringMathTex(
  File "/Users/behackl/code/manim/manim/mobject/svg/tex_mobject.py", line 260, in __init__
    file_name = tex_to_svg_file(
  File "/Users/behackl/code/manim/manim/utils/tex_file_writing.py", line 44, in tex_to_svg_file
    dvi_file = compile_tex(
  File "/Users/behackl/code/manim/manim/utils/tex_file_writing.py", line 191, in compile_tex
    raise ValueError(
ValueError: latex error converting to dvi. See log output above or the log file: media/Tex/9bb84ea3b159150a.log
```

I'd like to get this PR merged so that the new behavior can be tested for a bit before the next release.

**Comment by jsonvillanueva** [comment]
> I don't think your example would break: the internal implementation balances brackets before compiling, and here it should be able to repair all of the problems.

I actually checked out the branch and printed out the return value --the split was as I expected:
image

But I guess I'm not really sure what nesting achieve. Might be worth mentioning that nesting doesn't make sense in the docstring... but lets move this forward. I do want to fix the formatting of the logging.warning with dedent before this though as the error comes out very strange looking. I'll commit that soon:tm:

**Review by jsonvillanueva — APPROVED** [review]

File: manim/mobject/svg/tex_mobject.py [hunk 1]
```diff
@@ -473,7 +479,9 @@ def break_up_tex_strings(self, tex_strings):
                 pieces.extend(re.split(pattern, s))
         else:
             pieces = tex_strings
-        return [p for p in pieces if p]
+        ret = [p for p in pieces if p]
+        print(ret)
+        return ret
```
Comment (behackl) [inline manim/mobject/svg/tex_mobject.py:484]: I don't think that this change here is intentional.

Comment (jsonvillanueva) [reply manim/mobject/svg/tex_mobject.py:484]: Yikes! I thought I reverted that selection but I guess I forgot to save changes before staging...

**Pull request merged**